Make ResourceManager.BaseName work as documented.

madelson opened the pull request:
This property is documented to return the "qualified namespace name and the root resource name of a resource file". However, previously when constructing a ResourceManager with a Type the BaseName property would instead return the name of the type without its namespace.

fix #74918

madelson committed:
Make ResourceManager.BaseName work as documented.

This property is documented to return the "qualified namespace name and
the root resource name of a resource file". However, previously when
constructing a ResourceManager with a Type the BaseName property would
instead return the name of the type without its namespace.

fix #74918

madelson committed:
Remove _locationInfo field

madelson commented on src/libraries/System.Private.CoreLib/src/System/Resources/ResourceManager.cs:
```diff
@@ -227,9 +226,13 @@ public ResourceManager(Type resourceSource)
+            char c = Type.Delimiter;
+            BaseNameField = nameSpace is null
+                ? resourceSource.Name
+                : string.Concat(nameSpace, new ReadOnlySpan<char>(in c), resourceSource.Name);
```
This is the key change. Previously, we'd store the type and set `BaseNameField` to be the type name. Then when looking up resources we'd pass in both the type and the name and `GetManifestResourceStream` would recombine them. Now, we just compute the correct base name upfront, which means we don't need to store the type at all.

One downside to this approach is that the change is observable to someone who is deriving from `ResourceManager` and modifying the `BaseNameField` field which is `protected`. However, I can't imagine a realistic use-case for that; having `BaseName` return the right/consistent thing in all cases feels more valuable.

buyaa-n replied:
Why could not use just the `FullName`? i.e. `BaseNameField = resourceSource.FullName!`

madelson replied:
Quite possibly we could do that and I'm happy to make the change. The reason I didn't do that initially is that the old code used this logic rather than `FullName` ; I thought there might be a reason for that.

buyaa-n replied:
Though in that case the resource name populated for generic types will be different:
- `FullName`: System.Collections.Generic.List`1[[System.String, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]
- `Namespace + '.' + Name`: System.Collections.Generic.List`1

Not really sure we should cover this scenario

tarekgh replied:
What was the old behavior when using the generic types? I prefer to reduce any breaking change here.

buyaa-n replied:
By old behavior the resource name will be "System.Collections.Generic.List`1"

joperezr commented on src/libraries/System.Private.CoreLib/src/System/Resources/FileBasedResourceGroveler.cs:
```diff
@@ -47,10 +47,7 @@ public FileBasedResourceGroveler(ResourceManager.ResourceManagerMediator mediato
-                        // expect the neutral locale's resources to be present.
-                        string? locationInfo = _mediator.LocationInfo == null ? "<null>" : _mediator.LocationInfo.FullName;
-                        throw new MissingManifestResourceException($"{SR.MissingManifestResource_NoNeutralDisk}{Environment.NewLineConst}baseName: {_mediator.BaseNameField}  locationInfo: {locationInfo}  fileName: {_mediator.GetResourceFileName(culture)}");
+                        throw new MissingManifestResourceException($"{SR.MissingManifestResource_NoNeutralDisk}{Environment.NewLineConst}baseName: {_mediator.BaseNameField}  fileName: {_mediator.GetResourceFileName(culture)}");
```
I'm not sure I understand how this change is related to the issue being fixed.

madelson replied:
We no longer store or track `_locationInfo`, seeing as it is now purely redundant with `fileName`

joperezr commented on src/libraries/System.Private.CoreLib/src/System/Resources/ManifestBasedResourceGroveler.cs:
```diff
@@ -321,43 +321,36 @@ private static ResourceSet InternalGetResourceSetFromSerializedData(Stream store
         }
 
-        private Stream? GetManifestResourceStream(Assembly satellite, string fileName)
+        private static Stream? GetManifestResourceStream(Assembly satellite, string fileName)
```
Same here, not sure how this relates to the BaseName change.

madelson replied:
Previously, this couldn't be static because it accessed `_mediator`, but that's no longer true since `fileName` now contains the full information in all cases. With that change, I believe I was getting an analyzer error until I made this static.

joperezr commented on src/libraries/System.Resources.ResourceManager/tests/ResourceManagerTests.cs:
```diff
@@ -204,7 +204,10 @@ public static void UsingResourceSet()
+            var manager = new ResourceManager("System.Resources.Tests.Resources.TestResx", typeof(ResourceManagerTests).Assembly);
+            Assert.Equal("System.Resources.Tests.Resources.TestResx", manager.BaseName);
+
+            manager = new(typeof(System.Resources.Tests.Resources.TestResx));
```
NIT: we only use the compressed new statement when the type can be inferred from the left side of the assignement which isn't the case here (even though it is two lines above).

joperezr commented:
Sorry for taking a while to look at this, thanks for the contribution!

Could you also please add more tests that ensure the right behavior of BaseName? For example, can you add one where the type doesn't have a namespace too?

madelson committed:
Add additional test cases for ResourceManager.BaseName.

Address feedback from https://github.com/dotnet/runtime/pull/75497#issuecomment-1262913704.

madelson commented on src/libraries/System.Resources.ResourceManager/tests/ResourceManagerTests.cs:
```diff
@@ -204,8 +204,19 @@ public static void UsingResourceSet()
+
+            manager = new ResourceManager(typeof(List<string>));
+            Assert.Equal("System.Collections.Generic.List`1", manager.BaseName);
         }
```
@joperezr any other test cases you'd like to see?

madelson commented on src/libraries/System.Resources.ResourceManager/tests/ResourceManagerTests.cs:
```diff
@@ -204,8 +204,19 @@ public static void UsingResourceSet()
+            Assert.Equal(typeWithoutNamespace.Name, manager.BaseName);
+
+            manager = new ResourceManager(typeof(List<string>));
+            Assert.Equal("System.Collections.Generic.List`1", manager.BaseName);
```
@joperezr This result is kind of interesting. We might expect `String` to be in there somewhere but with the current and previous logic for computing type names the generic parameters don't matter.

buyaa-n replied:
Is this really a valid case to test? When a generic type could be used?

madelson replied:
I feel like it is equally valid as any other "irregular" test case (such as testing a type without a namespace). `ResourceManager`'s `Type` constructor allows any type and just looks for resources with the same full name in that type's assembly; AFAIK there are no limitations on what resource files can be named and you don't have to use the built-in generator.

That said, if you don't want this test I'm happy to remove it.

buyaa-n replied:
As you know the base name is a physical file name without the extension. Even if technically the argument passed could be a generic type and the resource file could be named "System.Collections.Generic.List`1", I don't think it is practical. I think it is uncovered edge case scenario from the beginning and doesn't feel like it's important scenario need to be cover now, if need to cover it should behave same as how it was working before instead of we making new rule for this now, what you think?

buyaa-n replied:
I think the current test is match with the old behavior, so resolving the comment

buyaa-n commented:
Thanks for the fix @madelson, but it might be not a bug needs a fix, see [my comment on the issue for details](https://github.com/dotnet/runtime/issues/74918#issuecomment-1279611539), we might need to clarify the doc instead, I will dig into it more

Also there is existing test that specifically checking this scenario is [failing on CI](https://helix.dot.net/api/2019-06-17/jobs/141cc7ff-dab1-44d0-b4e1-5a4f5429cbcd/workitems/System.ComponentModel.TypeConverter.Tests/console)

```log
 System.ComponentModel.Tests.ComponentResourceManagerTests.Ctor_Type(type: typeof(int)) [FAIL]
      Assert.Equal() Failure
                ↓ (pos 0)
      Expected: Int32
      Actual:   System.Int32
                ↑ (pos 0)
      Stack Trace:
        /_/src/libraries/System.ComponentModel.TypeConverter/tests/ComponentResourceManagerTests.cs(27,0): at System.ComponentModel.Tests.ComponentResourceManagerTests.Ctor_Type(Type type)
           at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
        /_/src/libraries/System.Private.CoreLib/src/System/Reflection/MethodInvoker.cs(69,0): at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
  Finished:    System.ComponentModel.TypeConverter.Tests
=== TEST EXECUTION SUMMARY ===
   System.ComponentModel.TypeConverter.Tests  Total: 7825, Errors: 0, Failed: 1, Skipped: 0, Time: 32.183s
```

madelson commented:
Thanks for engaging @buyaa-n . I'll fix that test if we decide to proceed with this change.

buyaa-n commented:
> Thanks for engaging @buyaa-n . I'll fix that test if we decide to proceed with this change.

Please do that so that we can see if any other test fails

madelson committed:
Fix BaseName test case in response to https://github.com/dotnet/runtime/pull/75497#issuecomment-1295500698

madelson commented:
> Please do that so that we can see if any other test fails

@buyaa-n pushed the fix!

buyaa-n reviewed (approved):
Overall LGTM, thanks @madelson
